thomwolf/transformers
# Bigscience176b

# What does this PR do?

Opening a draft pull request for:

## Testing the BigScience176B tokenizer

To test the tokenizer we:
+ Generate tokens from a dummy text. Encode it and decode it back
+ Make sure that the tokens are the same than the one created by megatron-deepspeed
+ For now (before the tokenizer 0.12.1 release) follow [these instructions](https://huggingface.co/bigscience-catalogue-data-dev/byte-level-bpe-tokenizer-no-norm-250k-whitespace-and-eos-regex-alpha-v3-dedup-lines-articles) to download and use the tokenizer.
+ Reconstruction tests on ```xnli``` dataset

### TO DO:

- [x] Push the .bin data in the Hub to be able to download it in local and use it? OR hardcode the values of the tokens (better)
- [ ] Discuss about slow tests
- [ ] Hardcoding tokens from ```xnli``` dataset

## Testing the embeddings from the converted model

I first:
+ Retrieved the embeddings generated from the GPT2 model that has been trained using Megatron-DeepSpeed. For now the mean value of each embedding is hardcoded.
Then:
+ Load the converted model and compare the values of the mean embeddings using the same input. The check passes **if and only if** the weights of the model class modules are with the same ```dtype``` as the ones computed on the script on Megatron-DeepSpeed. With this implementation should be able to try the test with ```float16```,  ```float32``` or ```bfloat16```and I hardcoded the mean values of the embeddings for each ```dtype```. 

### What do we needed to make it work?

+ Add a specific dtype when loading the module [779bdd0](https://github.com/thomwolf/transformers/pull/1/commits/779bdd09bc3bfbfefcf4cb96baa5a500992ab60d) & [4f17f81](https://github.com/thomwolf/transformers/pull/1/commits/4f17f8151f9c6e6389eb9a9149a8aee95cc805f9)

### TO DO:

The test **passes** for the logits before the layer norm for all supported ```dtype```, but not passing for the logits after the layernorm, I suspect that:
- [x] The weights were not loaded correctly into ```bfloat16``` (exactly [here](https://github.com/thomwolf/transformers/blob/88f05f40100ae9e2f94bacfca5302c2dd77b0ab1/src/transformers/models/bigscience176b/convert_bigscience176b_original_checkpoint_to_pytorch.py#L107), but even after dividing by an ```int```, the weights are still ```bfloat16```)
- [x] We have to implement the ```FusedLayerNorm``` on the model definition, since this is used during training.
- [x] Add more metrics (max, min, median)?
- [x]  Understand why we do not obtain the same values on GPU and CPU

More exciting things to come!

 - [x] Make sure that apex is **not installed** in the virtual env that I am using
 - [x] Check the vales of the attention layer (before the MLP)
 - [x] Figure out how to merge the different outputs across TP ranks -> To discuss with Thomas just to be sure
 - [x] Add biases on each MLP (should work)
 - [x] Check ```dense_4h_to_h```?
 - [x] Check the values of the logits after the MLP
 - [x] Get the final logits
